llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,319 @@
+// from parallel topk and computing the final combined result.
+iree_compiler::IREE::LinalgExt::TopkOp
+computeReductionTopk(Location loc, PatternRewriter &rewriter,
+                     iree_compiler::IREE::LinalgExt::TopkOp topkOp,
+                     iree_compiler::IREE::LinalgExt::TopkOp parallelTopkOp,
+                     Value updatedParallelIndices,
+                     ArrayRef<ReassociationIndices> reassociationIndices,
+                     int64_t splitReductionRatio, int64_t kDimOrig,
+                     int64_t kSize) {
+  Value valuesOrig = topkOp.values();
+  ShapedType valuesOrigType = valuesOrig.getType().cast<ShapedType>();
+  Type valueElementType = valuesOrigType.getElementType();
+  Type indicesElementType =
+      topkOp.getResultTypes()[1].cast<ShapedType>().getElementType();
+
+  // Define the collapsed input shapes
+  SmallVector<int64_t> collapsedShape = getCollapsedShape(
+      valuesOrigType.getShape(), splitReductionRatio, kSize, kDimOrig);
+  RankedTensorType valuesCollapsedType =
+      RankedTensorType::get(collapsedShape, valueElementType);
+  RankedTensorType indicesCollapsedType =
+      RankedTensorType::get(collapsedShape, indicesElementType);
+
+  // Collapse collapse parallel output for the input of final reduction
+  Value valuesCollapsed = rewriter.create<tensor::CollapseShapeOp>(
+      loc, valuesCollapsedType, parallelTopkOp.getResults()[0],
+      reassociationIndices);
+  Value indicesCollapsed = rewriter.create<tensor::CollapseShapeOp>(
+      loc, indicesCollapsedType, updatedParallelIndices, reassociationIndices);
+
+  // Combined final topk
+  auto reductionTopkOp =
+      rewriter.create<iree_compiler::IREE::LinalgExt::TopkOp>(
+          loc,
+          /* resultTypes= */ topkOp->getResultTypes(),
+          /* ins= */ ValueRange{valuesCollapsed, indicesCollapsed},
+          /* outs= */ topkOp.outputs(), kDimOrig);
+  rewriter.cloneRegionBefore(topkOp.region(), reductionTopkOp.region(),
+                             reductionTopkOp.region().end());
+  return reductionTopkOp;
+}
+
+} // namespace
+
+// Transforms an applicable standard single reduction TopkOp into a parallel
+// reduction TopkOp with a reduce step following.
+//
+// Handles parallel reductions in 2 phases: A "map" parallel phase and the a
+// single "reduce" reduction phase. The first phase expands the input tensor
+// shape by breaking the reduction dimension into multiple parallel reductions
+// (upping the rank of the input). Topk is run on these dimensions in parallel
+// The second phase collapses the parallel results into a single final reduce.
+// Topk is run again on the combined output to produce a final output.
+//
+// Currently only topk operations without input indices are supported.
+FailureOr<TopkOp> mlir::iree_compiler::IREE::LinalgExt::TopkOpSplitReduction::
+    returningMatchAndRewrite(iree_compiler::IREE::LinalgExt::TopkOp topkOp,
```

Comment:
I mentioned this previously too, this is not the expected way, and there is no reason to use `returningMatchAndRewrite`. For now just use `matchAndRewrite`.
In any case, `returningMatchAndRewrite` is expected to have the same signature as `matchAndRewrite`.